walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -7,17 +7,105 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
 package com.google.sps;
 
+import java.util.Queue;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Comparator;
+import java.util.ArrayList;
+import java.util.List;
+
+/* A meeting request has:
+   - a name
+   - a duration in minutes
+   - a collection of attendees
+   
+   For a particular time slot to work, all attendees must be free to attend the meeting. When a query is made, it will be given a collection of all known events. Each event has:
+   - a name
+   - a time range
+   - a collection of attendees
+   A time range will give you the start time, the end time, and the duration of the event. If you want to know more, open the TimeRange.java file.
+*/
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Creates an ArrayList to hold busy times
+    List<TimeRange> busyTimes = new ArrayList<>();
+
+    // Gets info of request
+    Collection<String> attendees = request.getAttendees();
+    long dur = request.getDuration();
+
+    // Special case where there can be no meetings of that duration 
+    if(dur > TimeRange.END_OF_DAY - TimeRange.START_OF_DAY || dur < 0){
+      return busyTimes; 
+    }
+    
+    // looks at every event
+    for(Event event : events){ 
+      Collection<String> eventAtt = event.getAttendees();
+      // looks at all the people attending the event
+      for(String attendee : eventAtt){ 
+        if(attendees.contains(attendee)){
+          busyTimes.add(event.getWhen());
+          break; // found someone in this event, we can stop searching this one
+        }
+      }
+    }
+    List<TimeRange> timeRangesAvail = new ArrayList(); 
+
+    // Assume whole day is free at first
+    timeRangesAvail.add(TimeRange.WHOLE_DAY);
+
+    // Sort the busy times by start date for easier parsing
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START); 
+
+    // Goes through each busyTime
+    for(TimeRange busyTime : busyTimes){
+      int origLen = timeRangesAvail.size();
+      for(int index = 0; index < origLen; index ++){
+        TimeRange freeTime = timeRangesAvail.get(index);
+        // remove fully, split free by busyTime
+        if(timeRangesAvail.get(index).contains(busyTime)){ 
+          TimeRange secHalf = TimeRange.fromStartDuration(busyTime.end(), freeTime.end() - busyTime.end());
+          TimeRange firstHalf = TimeRange.fromStartDuration(freeTime.start(), freeTime.duration() - secHalf.duration() - busyTime.duration());
+          timeRangesAvail.remove(freeTime); // remove old time
+          index--; // back the counter up because we removed an item
```

Comment:
Reading this code got me curious about whether there's a way to iterate over a `List` while being able to add/remove elements from it. I found the `ListIterator` class: https://docs.oracle.com/javase/8/docs/api/java/util/ListIterator.html

I don't know if that would be helpful here, but it might be worth checking out.